pkg/operator/ceph/cluster/cluster.go

```diff
@@ -397,6 +411,86 @@ func (c *ClusterController) preClusterStartValidation(cluster *cluster, clusterO
 	return nil
 }
 
+func extractExitCode(err error) (int, bool) {
+	exitErr, ok := err.(*exec.ExitError)
+	if ok {
+		return exitErr.ExitCode(), true
+	}
+	return 0, false
+}
+
+func (c *cluster) removeDefaultReplicationRule() error {
+	args := []string{"osd", "crush", "rule", "rm", "replicated_rule"}
+	cephCmd := client.NewCephCommand(c.context, c.Namespace, args)
+	_, err := cephCmd.Run()
+	if err != nil {
+		if code, ok := extractExitCode(err); ok && code == int(syscall.EBUSY) {
+			// we do not want to delete the replicated_rule if it’s in use,
+			// and we also do not care much. There are two possible causes:
+			// - the user has created this rule with the non-default CRUSH
+			//   root manually
+			// - the user is using this rule despite the rule using the default
+			//   CRUSH root
+			// in both cases, we cannot do anything about it either way and
+			// we’ll assume that the user knows what they’re doing.
+			logger.Debug("replicated_rule is in use, not removed")
+			return nil
+		}
+		// the error does not refer to EBUSY -> return as error
+		return errors.Wrap(err, "failed to remove default replicated_rule")
+	}
+	return nil
+}
+
+func (c *cluster) removeDefaultCrushRoot() error {
+	args := []string{"osd", "crush", "rm", "default"}
+	cephCmd := client.NewCephCommand(c.context, c.Namespace, args)
+	_, err := cephCmd.Run()
+	if err != nil {
+		if code, ok := extractExitCode(err); ok {
+			if code == int(syscall.ENOTEMPTY) || code == int(syscall.EBUSY) {
+				// we do not want to delete the default node if it’s in use,
+				// and we also do not care much. There are two more causes here:
+				// - a (non-root?) CRUSH node with the default label was created
+				//   automatically, e.g. from topology labels, and OSDs (or sub
+				//   nodes) have been placed in there. In this case, the node
+				//   obviously needs to be preserved.
+				// - the root=default CRUSH node is in use by a non-default
+				//   CRUSH rule
+				// - OSDs or subnodes have been placed under the root=default
+				//   CRUSH node
+				//
+				// in all cases, we cannot do anything about it either way and
+				// we’ll assume that the user knows what they’re doing.
+				logger.Debug("default is not empty or is still in use, not removed")
```

Comment:
@leseb Same as above. This message will *always* appear in a cluster which uses a non-default root for a good reason, since the `osd crush rm` command only goes by "value" and not by "level": We can’t say `osd crush rm root=default`, but can only say `osd crush rm default`.

That means that when the non-default root is chosen because, e.g. the region topology key is `default`, every cluster operation will cause this message to be logged; the operator will execute this and fail to `osd crush rm default`, because the (e.g.) `region=default` CRUSH node is not empty or still in use.

An error would be even more confusing, IMO, than in the other case.

Sorry if this was obvious. I just want to make sure we’re on the same page.